cmd/mt-index-migrate/main.go

```diff
@@ -0,0 +1,167 @@
+		log.Fatal(4, "failed to create cql session. %s", err)
+	}
+
+	// ensure the new table exists.
+	err = session.Query(fmt.Sprintf(table_schema, *keyspace)).Exec()
+	if err != nil {
+		log.Fatal(4, "cassandra-idx failed to initialize cassandra table. %s", err)
+	}
+	err = session.Query(fmt.Sprintf(metric_idx_index, *keyspace)).Exec()
+	if err != nil {
+		log.Fatal(4, "cassandra-idx failed to initialize cassandra index. %s", err)
+	}
+
+	wg.Add(1)
+	go writeDefs(session, defsChan)
+	wg.Add(1)
+	go getDefs(session, defsChan)
+
+	wg.Wait()
+
+}
+
+func writeDefs(session *gocql.Session, defsChan chan *schema.MetricDefinition) {
+	log.Info("starting write thread")
+	defer wg.Done()
+	counter := 0
+	pre := time.Now()
+	for def := range defsChan {
+		qry := `INSERT INTO metric_idx (id, orgid, partition, name, metric, interval, unit, mtype, tags, lastupdate) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`
+		if *dryRun {
+			fmt.Printf(
+				"INSERT INTO metric_idx (id, orgid, partition, name, metric, interval, unit, mtype, tags, lastupdate) VALUES ('%s', '%d', '%d','%s', '%s','%d', '%s','%s', '%v', '%d')\n",
+				def.Id,
+				def.OrgId,
+				def.Partition,
+				def.Name,
+				def.Metric,
+				def.Interval,
+				def.Unit,
+				def.Mtype,
+				def.Tags,
+				def.LastUpdate)
+			continue
+		}
+		success := false
+		attempts := 0
+		for !success {
+			if err := session.Query(
+				qry,
+				def.Id,
+				def.OrgId,
+				def.Partition,
+				def.Name,
+				def.Metric,
+				def.Interval,
+				def.Unit,
+				def.Mtype,
+				def.Tags,
+				def.LastUpdate).Exec(); err != nil {
+
+				if (attempts % 20) == 0 {
+					log.Warn("cassandra-idx Failed to write def to cassandra. it will be retried. %s", err)
+				}
+				sleepTime := 100 * attempts
+				if sleepTime > 2000 {
+					sleepTime = 2000
+				}
+				time.Sleep(time.Duration(sleepTime) * time.Millisecond)
+				attempts++
+			} else {
+				success = true
+				log.Debug("cassandra-idx metricDef saved to cassandra. %s", def.Id)
+				counter++
+			}
+		}
+	}
+	log.Info("Inserted %d metricDefs in %s", counter, time.Since(pre).String())
+}
+
+func getDefs(session *gocql.Session, defsChan chan *schema.MetricDefinition) {
+	log.Info("starting read thread")
+	defer wg.Done()
+	defer close(defsChan)
+	partitioner, err := cluster.NewKafkaPartitioner(*partitionBy)
+	if err != nil {
+		log.Fatal(4, "failed to initialize partitioner. %s", err)
+	}
+	iter := session.Query("SELECT def from metric_def_idx").Iter()
+
+	var data []byte
+	for iter.Scan(&data) {
+		mdef := schema.MetricDefinition{}
+		_, err := mdef.UnmarshalMsg(data)
+		if err != nil {
+			log.Error(3, "cassandra-idx Bad definition in index. %s - %s", data, err)
+			continue
+		}
+		log.Debug("retrieved %s from old index.", mdef.Id)
+		if *numPartitions == 1 {
+			mdef.Partition = 0
+		} else {
+			p, err := partitioner.Partition(&mdef, int32(*numPartitions))
+			if err != nil {
+				log.Error(3, "failed to get partition id of metric. %s", err)
+				mdef.Partition = 0
```

Comment:
This condition seems like a big problem warranting an abort instead of proceeding with 0 partition